tests/src/Interop/StringMarshalling/Native/StringMarshalingNative.h

```diff
@@ -0,0 +1,109 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include <platformdefines.h>
+#include <type_traits>
+#include <algorithm>
+
+template<typename StringT>
+struct StringTraits
+{
+    using type = StringT;
+    using CharT = typename std::remove_pointer<StringT>::type;
+    using ConstStringT = CharT const *;
+    static const bool VerifyNullTerminator = true;
+};
+
+struct BSTRTraits
+{
+    using type = BSTR;
+    using ConstStringT = BSTR;
+    static const bool VerifyNullTerminator = false;
+};
+
+template<typename StringT>
+using VerifyReversedCallback = BOOL (__cdecl*)(StringT original, StringT reversed);
+
+template<typename StringT>
+using ReverseCallback = void(__cdecl*)(StringT original, StringT* reversed);
+
+template<typename StringT>
+using ReverseCallback = void(__cdecl*)(StringT original, StringT* reversed);
+
+template<typename StringT>
+using ReverseCallbackReturned = StringT(__cdecl*)(StringT original);
+
+template<typename StringT>
+using ReverseInplaceCallback = void(__cdecl*)(StringT str);
+
+template<typename StringTraitsT, size_t LengthFunction(typename StringTraitsT::ConstStringT), typename CharT = typename StringTraitsT::CharT>
+struct StringMarshalingTestsBase
+{
+    using StringT = typename StringTraitsT::type;
+    using ConstStringT = typename StringTraitsT::ConstStringT;
+
+    static BOOL Compare(ConstStringT expected, const StringT actual)
+    {
+        if (LengthFunction(expected) != LengthFunction(actual))
+        {
+            return FALSE;
+        }
+
+        size_t length = LengthFunction(expected);
+
+        for (size_t i = 0; i < length + (StringTraitsT::VerifyNullTerminator ? 1 : 0); i++)
```

Comment:
This is the classic loop for compare and it works perfectly. I would instead leverage the iterator approach since it is much easier to debug the failing condition.

``` C++
size_t length = LengthFunction(expected) + (StringTraitsT::VerifyNullTerminator ? 1 : 0);
ConstStringT currE = expected;
ConstStringT endE = expected + length;
const StringT currA = actual;
const StringT endA = actual + length;
for (; currE != endE; ++currE, ++currA)
{
   if (((CharT)*currE) != ((CharT)*currA))
       return FALSE;
}
```